streams/src/main/java/org/apache/kafka/streams/state/internals/NamedCache.java

```diff
@@ -200,6 +201,7 @@ synchronized void evict() {
         }
         remove(eldest);
         cache.remove(eldest.key);
+        dirtyKeys.remove(eldest.key);
```

Comment:
Another observation I had is that in `CachingKeyValueStore`, we set the listener as

```
                final List<KeyValue<Bytes, byte[]>> keyValues = new ArrayList<>();
                for (ThreadCache.DirtyEntry entry : entries) {
                    keyValues.add(KeyValue.pair(entry.key(), entry.newValue()));
                    maybeForward(entry, (InternalProcessorContext) context);
                }
                underlying.putAll(keyValues);
```

We are re-ordering the operation to first-forward-then-store, whereas in the old code we used to store-first-then-forward. Would that have any potential issue?